7989-refactor-preview: Support for Refactor Preview functionality

danarad05 opened the pull request:
Signed-off-by: Dan Arad <dan.arad@sap.com>

#### What it does
Resolves #7989 - Adds Refactor Preview functionality.

Remarks:
These features existing in vscode were not implemented currently in this PR:
1. Checkboxes for selecting which changes to implement
2. Upon right clicking on text line in Refactor Preview pane - context menu is not displayed like in vscode.

CQ: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=23016

#### How to test
1. Open ts file, click F2 for renaming a function
2. Enter new name and press Shift+Enter
3. Refactor Preview pane should open in bottom with changes
4. Press V button on pane toolbar to accept and apply changes
5. Press X button to discard changes

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

vince-fugnitto reviewed (commented):
@danarad05 CI is [failing](https://travis-ci.com/github/eclipse-theia/theia/jobs/395061747#L8051) since the newly added extension does not include a test file such as:

https://github.com/eclipse-theia/theia/blob/master/packages/getting-started/src/package.spec.ts

danarad05 commented:
> @danarad05 CI is [failing](https://travis-ci.com/github/eclipse-theia/theia/jobs/395061747#L8051) since the newly added extension does not include a test file such as:
> 
> https://github.com/eclipse-theia/theia/blob/master/packages/getting-started/src/package.spec.ts

Thanks Vince, just saw it and in process of adding spec file. Thanks

vince-fugnitto commented on packages/bulk-edit/package.json:
```diff
@@ -0,0 +1,52 @@
+{
+  "name": "@theia/bulk-edit",
+  "version": "1.5.0",
```
The version should be updated to match the rest of the application, namely `1.6.0` at this time.

danarad05 replied:
Fixed. Thanks

vince-fugnitto commented on packages/bulk-edit/src/browser/style/bulk-edit.css:
```diff
@@ -0,0 +1,65 @@
+
+.theia-bulk-edit-container .bulkEditNode .message .replaced-text {
+    text-decoration: line-through;
+    background-color: #4B1818;
```
We should instead define variables for color, like 
```css
color: var(--theia-descriptionForeground);
```

So that it will work nicely for all themes. If there is no color available then we can include a new `ColorContribution` for this extension.

danarad05 replied:
@vince-fugnitto 
Tried to find where **--theia-descriptionForeground** is defined and can't seem to find it (I copied styling from callhierarchy extension). Could you direct me to it?

vince-fugnitto replied:
It should come directly from `monaco`, if the color does not exist for your use-case (similar to vscode), then a color contribution is required: https://github.com/microsoft/monaco-editor/blob/master/test/playground.generated/customizing-the-appearence-exposed-colors.html#L66

You can reference the color documentation when theming support was added: https://github.com/eclipse-theia/theia/pull/6475/files#diff-4ac32a78649ca5bdd8e0ba38b7006a1eR9

danarad05 replied:
Fixed. Thanks

vince-fugnitto commented on packages/bulk-edit/src/browser/style/bulk-edit.css:
```diff
@@ -0,0 +1,65 @@
+
+.theia-bulk-edit-container .bulkEditNode .message .inserted-text {
+    text-decoration: line-through;
+    background-color: #373D29;
```
Same as previous color comment.

danarad05 replied:
Fixed. Thanks

vince-fugnitto reviewed (commented):
I have a few general comments, I have not yet tested the functionality.

We should also add the `@theia/bulk-edit` extension to both examples (`browser` and `electron`) since we explicitly list all the available extensions in both examples for others to reference.

amiramw commented:
@danarad05 you should usually keep single commit in your PR

danarad05 commented:
Ok, I did create a new commit with last changes done but found I missed a change and also had to fix Vince’s change requests.

Just to understand better – is there any reasoning for this rule? Why should there be only one commit in PR?

Thanks

From: Amiram Wingarten <notifications@github.com>
Sent: Tuesday, October 6, 2020 1:05 AM
To: eclipse-theia/theia <theia@noreply.github.com>
Cc: Arad, Dan (external - Service) <dan.arad@sap.com>; Mention <mention@noreply.github.com>
Subject: Re: [eclipse-theia/theia] 7989-refactor-preview: Support for Refactor Preview functionality (#8589)


@danarad05<https://github.com/danarad05> you should usually keep single commit in your PR

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/eclipse-theia/theia/pull/8589#issuecomment-703914036>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AQ6G34X55L3LGVHTHFH5RBLSJI7HXANCNFSM4SEOTJPA>.

danarad05 commented:
@vince-fugnitto 

What does this error mean?
![image](https://user-images.githubusercontent.com/71069170/95296508-9182bd80-0881-11eb-83a3-45c9d4ae2bb4.png)

vince-fugnitto commented:
> @vince-fugnitto
> 
> What does this error mean?
> ![image](https://user-images.githubusercontent.com/71069170/95296508-9182bd80-0881-11eb-83a3-45c9d4ae2bb4.png)

I believe it's unrelated, I'll restart the build for macOS.

idoprz commented:
Hi Dan,

I tried to run one of our scenarios that uses it, and it didn't opened the preview.
Can you take a look at this:
https://github.com/SAP/code-snippet

I can help you run it if necessary.

Thanks,
Ido.

danarad05 commented:
@vince-fugnitto 
Finished with current PR. please review.
Thanks
Dan

vince-fugnitto commented:
> @vince-fugnitto
> Finished with current PR. please review.

@danarad05 I'd be happy to take a look. I have a few other reviews in my backlog along with EclipseCon this week so it may take some time before I am able to review it thoroughly.

vince-fugnitto commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx:
```diff
@@ -0,0 +1,202 @@
+        super(treeProps, model, contextMenuRenderer);
+
+        this.id = BULK_EDIT_TREE_WIDGET_ID;
+        this.title.label = 'Refator Preview';
```
```suggestion
        this.title.label = 'Refactor Preview';
```

danarad05 replied:
Fixed

vince-fugnitto commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx:
```diff
@@ -0,0 +1,202 @@
+
+        this.id = BULK_EDIT_TREE_WIDGET_ID;
+        this.title.label = 'Refator Preview';
+        this.title.caption = 'Refator Preview';
```
```suggestion
        this.title.caption = 'Refactor Preview';
```

danarad05 replied:
Fixed

vince-fugnitto commented on packages/bulk-edit/package.json:
```diff
@@ -0,0 +1,50 @@
+  "version": "1.6.0",
+  "description": "Theia - Bulk Edit Extension",
+  "dependencies": {
+    "@theia/core": "^1.6.0",
+    "@theia/editor": "^1.6.0",
+    "@theia/monaco": "^1.6.0",
+    "@types/p-queue": "^2.3.1",
+    "@theia/filesystem": "^1.6.0",
+    "@theia/navigator": "^1.6.0",
+    "@theia/workspace": "^1.6.0",
```
Please sort these dependencies (it is done automatically during releases but it will result in changes).

danarad05 replied:
Fixed

vince-fugnitto commented on packages/bulk-edit/src/browser/bulk-edit-contribution.ts:
```diff
@@ -0,0 +1,106 @@
+import { AbstractViewContribution } from '@theia/core/lib/browser/shell/view-contribution';
+import { BulkEditCommands } from './bulk-edit-commands';
+import { MonacoBulkEditService } from '@theia/monaco/lib/browser/monaco-bulk-edit-service';
+import { TabBarToolbarContribution, TabBarToolbarRegistry } from '@theia/core/src/browser/shell/tab-bar-toolbar';
```
```suggestion
import { TabBarToolbarContribution, TabBarToolbarRegistry } from '@theia/core/lib/browser/shell/tab-bar-toolbar';
```

danarad05 replied:
Fixed

vince-fugnitto commented on packages/bulk-edit/README.md:
```diff
@@ -0,0 +1,30 @@
+
+## Description
+
+The `@theia/bulk-edit` extension contributes an `bulk-edit` widget to the application that displays WorkspaceEdits to end-users.
```
Should we instead describe the contributed widget as the _Preview Refactor_?

```suggestion
The `@theia/bulk-edit` extension contributes an `preview refactor` widget to the application that displays WorkspaceEdits to end-users.
```

danarad05 replied:
Fixed

vince-fugnitto commented on packages/bulk-edit/src/browser/bulk-edit-contribution.ts:
```diff
@@ -0,0 +1,106 @@
+        return false;
+    }
+
+    private async _previewEdit(workspaceEdit: monaco.languages.WorkspaceEdit): Promise<monaco.languages.WorkspaceEdit> {
```
```suggestion
    private async previewEdit(workspaceEdit: monaco.languages.WorkspaceEdit): Promise<monaco.languages.WorkspaceEdit> {
```
https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#no_underscore_private

danarad05 replied:
Fixed

vince-fugnitto reviewed (commented):
I have a few general comments already regarding the extension (I have not looked in depth yet for the source-code):

- when the pull-request is built, it produces unstaged changes to `.travis.yml`:

```
evinfug@elx7849163z[±|danarad05-7989-refactor-preview U:1 ✗]:~/workspaces/theia $ git diff
diff --git a/.travis.yml b/.travis.yml
index 7e1f979e1..34b4cc8aa 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -19,7 +19,6 @@ cache:
     - examples/browser/node_modules
     - examples/electron/node_modules
     - node_modules
-    - packages/bulk-edit/node_modules
     - packages/callhierarchy/node_modules
     - packages/console/node_modules
     - packages/core/node_modules
```

- I believe the styling of the refactor (in the editor) should be adjusted (ex: black font), and the message is poorly formatted:

_theia_:

![refactor](https://user-images.githubusercontent.com/40359487/96767267-4cdf4200-13aa-11eb-9d39-66d7a959133a.gif)


_vscode_:

![bulk-vscode](https://user-images.githubusercontent.com/40359487/96768032-54531b00-13ab-11eb-8b91-1a6f9757a166.gif)

danarad05 commented:
@vince-fugnitto 
Concerning: https://github.com/eclipse-theia/theia/pull/8589#pullrequestreview-514061140
1. Fixed the font color.
2. In regard to the rename label - it is displayed correctly for me, I'm not sure why it is displayed like so for you. 
Seems like the keybindings are not populated for you? Do you have any idea why?
https://github.com/microsoft/vscode/blob/f74e473238aca7b79c08be761d99a0232838ca4c/src/vs/editor/contrib/rename/renameInputField.ts#L85

![image](https://user-images.githubusercontent.com/71069170/96857469-ae191b00-1467-11eb-941f-c96a30c9002e.png)

Thanks,
Dan

vince-fugnitto commented:
> **_when the pull-request is built, it produces unstaged changes to .travis.yml:_**
> Not sure what you mean exactly - travis.yml is in file changed in PR. Please elaborate.

Sure, I did the following:
1. checkout the branch using the command line instructions from GitHub:

```
git checkout -b danarad05-7989-refactor-preview master
git pull https://github.com/danarad05/theia.git 7989-refactor-preview
```

2. perform a `git clean -ffdx` 
3. perform a fresh build with `yarn`
4. when the build is completed, unstaged changes related to `travis.yml` are produced

![image](https://user-images.githubusercontent.com/40359487/96870218-91b4b880-143e-11eb-9402-15dac0c9c3e3.png)

Same thing happens when viewing the pull-request on Gitpod:

![image](https://user-images.githubusercontent.com/40359487/96872964-4d2b1c00-1442-11eb-9b90-08c923403594.png)

> In regard to the rename label - it is displayed correctly for me, I'm not sure why it is displayed like so for you.
Seems like the keybindings are not populated for you? (or they contain CRLF) Do you have any idea why?

I'm not sure why unfortunately, looking at the keyboard shortcuts I see that the keybinding is properly set.
For additional information I'm on Linux:

![image](https://user-images.githubusercontent.com/40359487/96871752-b14ce080-1440-11eb-861d-21a670fcf54b.png)

danarad05 commented:
mistakenly closed

RomanNikitenko reviewed (commented):
Tested the changes after today's update.
I see some differences from VS Code behavior, could you take a look:

Tab of an editor preview doesn't have title and the corresponding icon:

![empty_tab](https://user-images.githubusercontent.com/5676062/97174055-36662b80-179a-11eb-85df-7546789a19eb.png)

VS Code:
![VSCode_tab](https://user-images.githubusercontent.com/5676062/97174247-7decb780-179a-11eb-8fd0-fbbb5d09d9a9.png)

RomanNikitenko commented:
At pressing `Discard Refactoring` button VS Code closes `Refactor Preview` panel and preview editor as well.

![VSCode_cancel](https://user-images.githubusercontent.com/5676062/97175638-7dedb700-179c-11eb-9414-b718c11cc3a7.gif)

Theia behavior - preview editor is open

![Theia_cancel](https://user-images.githubusercontent.com/5676062/97175743-a37ac080-179c-11eb-9cbc-dcf6863e501d.gif)

RomanNikitenko commented:
Displaying of preview editor takes about 4-5 seconds for me.
Do you know why there is this delay and if it could be improved.

![Preview_delay](https://user-images.githubusercontent.com/5676062/97176638-e9845400-179d-11eb-8ed2-831261c73354.gif)

RomanNikitenko commented:
At clicking an item in the `Refactor Preview` panel incorrect line number is selected in an `refactor preview` editor. 
Is it related to your remark #2 in the PR description?

![line_number](https://user-images.githubusercontent.com/5676062/97178186-d1153900-179f-11eb-82bf-996f1edd9203.gif)

RomanNikitenko commented:
I see that in VS Code double click on items in `Refactor Preview` panel selects the corresponding line number in `refactor preview` editor.
Looks like Theia opens `refactor preview` editor + an editor.

VS Code

![VSCode_doubleClick](https://user-images.githubusercontent.com/5676062/97179245-46cdd480-17a1-11eb-985a-4ba6572c62b3.gif)

Theia

![Theia_doubleClick](https://user-images.githubusercontent.com/5676062/97180472-c60fd800-17a2-11eb-8553-a7c32587e837.gif)

RomanNikitenko commented:
**Minor**
VS Code allows to display `Refactor Preview` panel if refactoring is in progress, I'm not able to display the panel after canceling refactoring for example.
Theia displays empty panel with the corresponding message.

VS Code

![VSCode_open_panel](https://user-images.githubusercontent.com/5676062/97181097-957c6e00-17a3-11eb-830e-f8bf1b62a3ff.gif)

Theia 

![Theia_open_panel](https://user-images.githubusercontent.com/5676062/97181279-c2308580-17a3-11eb-861a-a7a8d4051928.gif)

danarad05 commented:
@RomanNikitenko 
@amiramw @DoroNahari 

Roman, 
First thanks for the code review.

Second, in regards to the issues you found:
**_1. Tab of an editor preview doesn't have title and the corresponding icon_**
> a) I'm using windows and no tabs show icon - I don't know if it's a bug or not - if it is a requirement then another issue should be created for it as it is not related to this current issue.
> b) In regard to sometimes not showing name - think it is fixed with my fix to issue 3 below.

**_2. On VSCODE at pressing Discard Refactoring button VS Code closes Refactor Preview panel and preview editor as well_**
> fixed

**_3. Displaying of preview editor takes about 4-5 seconds for me_**
> I used same solution for diff editor as exists in search and replace. Performance issue was created because in that solution
file contents containing the changes yet to be applied was concatenated to URI query which resulted in extreme performance setback as the file content becomes larger. Now after fix it is almost instantaneous. Problem was also fixed for search and replace.

**_4. At clicking an item in the Refactor Preview panel incorrect line number is selected in an refactor preview editor. Is it related to your remark 2 in the PR description_**
> fixed

**_5. double click on items in Refactor Preview panel selects the corresponding line number in refactor preview editor. Looks like Theia opens refactor preview editor + an editor _**
> a) In refactor preview - double click on item displays the file. one click on item displays diff editor. This is existing functionality - if this should be changed let me know.
> b) fixed correct line is selected 

**_6. VS Code allows to display Refactor Preview panel if refactoring is in progress, I'm not able to display the panel after canceling refactoring for example. Theia displays empty panel with the corresponding message:
> a) Current system functionality (as far as I can tell) is disposing of widget on closing of tab (on X button) - so I cannot display workspaceEdits after it.
> b) I fixed so that "Refactor Preview" in displayed in quick open view only when "Refactor Preview" bottom pane is displayed. otherwise it is hidden from quick open view.
When Refactor Preview pane is displayed:
![image](https://user-images.githubusercontent.com/71069170/97596877-decff600-1a0d-11eb-86bf-84c7ad69b1d3.png)

When Refactor Preview pane is not displayed:
![image](https://user-images.githubusercontent.com/71069170/97596841-d37cca80-1a0d-11eb-913b-a494c7bc8fca.png)

Please review again PR.

Thanks,
Dan

RomanNikitenko commented:
@danarad05 
> a) I'm using windows and no tabs show icon - I don't know if it's a bug or not - if it is a requirement then another issue should be created for it as it is not related to this current issue.

Could you check what's value you have for `File Icon Theme`

![FileIconTheme](https://user-images.githubusercontent.com/5676062/97856858-8b0f2680-1d05-11eb-8087-3be7daff6c50.gif)

danarad05 commented:
https://github.com/eclipse-theia/theia/pull/8589#issuecomment-720378344

@RomanNikitenko 
Regarding the icon theme - thanks for explanation - my bad - didn't use icons theme in vscode ever.
Anyway, I tried all options and icons are displayed for all icon themes for me currently - could you check again please? 

![image](https://user-images.githubusercontent.com/71069170/98086512-8f634d00-1e87-11eb-8bb8-263f087fbff8.png)

RomanNikitenko commented:
> could you check again please?

Will test today and let you know the result.

RomanNikitenko reviewed (commented):
Tested after the PR update:
> **1. Tab of an editor preview doesn't have title and the corresponding icon**

Can not reproduce it anymore. I still see the difference with VS Code for a tab of an editor preview, but it's OK for me - a title and an icon is displayed. 
 
> **2. At pressing Discard Refactoring button VS Code closes Refactor Preview panel and preview editor as well.**

Fixed.

> **3. Displaying of preview editor takes about 4-5 seconds for me**

Fixed - thanks a lot! it was really slow...

> **4. At clicking an item in the Refactor Preview panel incorrect line number is selected in an refactor preview editor. Is it related to your remark 2 in the PR description**

Still can reproduce it for first item in `Refactor Preview` panel.
To reproduce the problem, please pay attention that:
- first I selected first item in `Refactor Preview` panel - and selected line is incorrect
- then I selected any item - selected number is correct
- then I selected again first item - and now selected line is correct for the first item 

![select_line_preview](https://user-images.githubusercontent.com/5676062/98248841-2b6f8000-1f7e-11eb-90cf-5b5d7b89a7bd.gif)

> **5. double click on items in Refactor Preview panel selects the corresponding line number in refactor preview editor. Looks like Theia opens refactor preview editor + an editor**

As user I use single click for this case, so it's not critical for me - I leave it at other reviewers and community discretion.  

> **6. VS Code allows to display Refactor Preview panel if refactoring is in progress, I'm not able to display the panel after canceling refactoring for example. Theia displays empty panel with the corresponding message**

> a) Current system functionality (as far as I can tell) is disposing of widget on closing of tab (on X button) - so I cannot display workspaceEdits after it.
b) I fixed so that "Refactor Preview" in displayed in quick open view only when "Refactor Preview" bottom pane is displayed. otherwise it is hidden from quick open view.
When Refactor Preview pane is displayed:

If I understand correctly the new behavior after the PR update is: closing `Refactor Preview` panel means that refactoring is cancelled. So it does the same actions as `Discard Refactoring`. Correct? 
VS Code allows to close the panel, then open again and continue to work with refactoring. As for me - can be improved in a separate PR and within another issue.

So, could you check:
>  4. At clicking an item in the Refactor Preview panel incorrect line number is selected in an refactor preview editor.  

I described steps to reproduce above ^^

RomanNikitenko commented:
btw: after building the project with your changes I see
![travis](https://user-images.githubusercontent.com/5676062/98255826-83aa8000-1f86-11eb-9d47-f3426683edd0.png)

danarad05 commented:
@RomanNikitenko 
Thanks for review.
_4. At clicking an item in the Refactor Preview panel incorrect line number is selected in an refactor preview editor. Is it related to your remark 2 in the PR description_
fixed - please review again.
Also think I fixed the issue with travis.yml.

Thanks

RomanNikitenko commented:
@danarad05 
Thanks for the PR update, Dan!
I'm going test the changes today.

RomanNikitenko commented on packages/bulk-edit/src/browser/bulk-edit-tree/in-memory-text-resource.ts:
```diff
@@ -0,0 +1,40 @@
+
+import { injectable } from 'inversify';
+import URI from '@theia/core/lib/common/uri';
+import { Resource, ResourceResolver } from '@theia/core/src/common';
```
`src` => `lib`?

danarad05 replied:
fixed thanks

RomanNikitenko commented on packages/plugin/src/theia.d.ts:
```diff
@@ -7231,25 +7231,28 @@ declare module '@theia/plugin' {
+         * @param metadata Optional metadata for the entry.
          */
-        replace(uri: Uri, range: Range, newText: string): void;
+        replace(uri: Uri, range: Range, newText: string, metadata?: WorkspaceEditMetadata): void;
```
I didn't find 
`export interface WorkspaceEditMetadata`
within `theia.d.ts`

btw: I see that [VS Code uses](https://github.com/microsoft/vscode/blob/14767349ab8b56cdff3c015b6a7c6d24831c387d/src/vs/vscode.d.ts#L3070) WorkspaceEdit**Entry**Metadata

danarad05 replied:
fixed

RomanNikitenko commented:
> Also think I fixed the issue with travis.yml.

After the PR update I don't see `travis.yml` related problem, but there are 4 not staged files:

![not_staged_files](https://user-images.githubusercontent.com/5676062/98535383-b0b0a880-228e-11eb-9c36-229b33149a63.png)

Do you have the same behavior on your side after building the changes?

Also there is `no message` squashed commit, please fix it before merge.

RomanNikitenko commented:
@danarad05 
It's strange, but I can not start the application with your changes:
![start_theia_error](https://user-images.githubusercontent.com/5676062/98540364-81059e80-2296-11eb-9620-83e85fc3c8ba.png)

![theia_starting_error](https://user-images.githubusercontent.com/5676062/98540388-85ca5280-2296-11eb-9d64-e4517674bdc5.gif)

Could someone check it: maybe some problems on my side, although I'm able to start `Theia` from `master` branch...

danarad05 commented:
https://github.com/eclipse-theia/theia/pull/8589#issuecomment-723952542
https://github.com/eclipse-theia/theia/pull/8589#issuecomment-723980023

@RomanNikitenko Thanks for review Roman, fixed all pending issues. please review again.

danarad05 commented:
@vince-fugnitto 
Can you still reproduce this issue?  
https://user-images.githubusercontent.com/40359487/96767267-4cdf4200-13aa-11eb-9d39-66d7a959133a.gif

vince-fugnitto commented:
> @vince-fugnitto
> Can you still reproduce this issue?
> https://user-images.githubusercontent.com/40359487/96767267-4cdf4200-13aa-11eb-9d39-66d7a959133a.gif

I still see the following issue: (Linux)

_Browser_:

![f2-browser](https://user-images.githubusercontent.com/40359487/98682598-1a35c300-2332-11eb-8544-d5f98c5cc4c8.gif)

_Electron_:

![f2-electron](https://user-images.githubusercontent.com/40359487/98682611-1e61e080-2332-11eb-9e1f-9b1220644aff.gif)

RomanNikitenko reviewed (approved):
I didn't compare VS Code implementation with the current changes, but tested after the PR update - the issues which I found within previous reviews are fixed.

There are still some differences with VS Code behavior, like:
- VS Code allows to close the `Refactor Preview` panel, then open again and continue to work with refactoring. 
- the `Refactor Preview` panel of VS Code contains `Maximize/Restore Panel Size` buttons
- the difference related to double click on items of `Refactor Preview` panel 
- the features defined in `Remarks` of the PR description 

But as for me - it can be improved in a separate PR within another issue. 

@danarad05 
please squash/remove commits like `7989-refactor-preview: updated to 1.7.0` before merge for pure git history.

RomanNikitenko commented:
@vince-fugnitto 
Sorry - I didn't catch what's the difference with styles you mean.
[Here](https://github.com/eclipse-theia/theia/pull/8589#pullrequestreview-514061140) I see black font, but [here](https://github.com/eclipse-theia/theia/pull/8589#issuecomment-724715487) it's not.

Could you clarify what's exactly you mean - I could check on my side as well...

vince-fugnitto commented:
> Could you clarify what's exactly you mean - I could check on my side as well...

@RomanNikitenko in my case, on `electron` the keybinding is not properly displayed at the bottom of the input (tested on Linux):


_Browser_:

![f2-browser](https://user-images.githubusercontent.com/40359487/98682598-1a35c300-2332-11eb-8544-d5f98c5cc4c8.gif)

_Electron_:

![f2-electron](https://user-images.githubusercontent.com/40359487/98682611-1e61e080-2332-11eb-9e1f-9b1220644aff.gif)

danarad05 commented:
> @vince-fugnitto
> Sorry - I didn't catch what's the difference with styles you mean.

@RomanNikitenko There is a pending issue concerning electron version on Linux - the label underneath textbox is missing and crooked.  We are having a problem reproducing this issue as we have no Linux machine.. trying to find a solution for that at the moment.

RomanNikitenko commented:
@vince-fugnitto @danarad05 
I see, thank you.

I can reproduce it for electron assembly as well using `Fedora`.
![electron_refact_bug](https://user-images.githubusercontent.com/5676062/98950383-98948f80-2501-11eb-91c7-ff6b9358bbae.jpg)

danarad05 commented:
> please squash/remove commits like `7989-refactor-preview: updated to 1.7.0` before merge for pure git history.
@RomanNikitenko Done. Thanks

danarad05 commented:
> @vince-fugnitto @danarad05
> I see, thank you.
> 
> I can reproduce it for electron assembly as well using `Fedora`.
> ![electron_refact_bug](https://user-images.githubusercontent.com/5676062/98950383-98948f80-2501-11eb-91c7-ff6b9358bbae.jpg)

Are you on Linux?

RomanNikitenko commented:
> Are you on Linux?

yes

amiramw commented on packages/bulk-edit/src/browser/bulk-edit-tree/in-memory-text-resource.ts:
```diff
@@ -0,0 +1,40 @@
+/********************************************************************************
+ * Copyright (C) 2018 TypeFox and others.
```
should it be SAP?

danarad05 replied:
I copied it (with a little change) from packages\search-in-workspace\src\browser\in-memory-text-resource.ts so that's why I left it like that. Should I change it?

danarad05 replied:
Fixed. thanks

amiramw commented on packages/bulk-edit/src/browser/bulk-edit-tree/in-memory-text-resource.ts:
```diff
@@ -0,0 +1,40 @@
+
+@injectable()
+export class InMemoryTextResourceResolver implements ResourceResolver {
+    resolve(uri: URI): Resource | Promise<Resource> {
```
you can use `MaybePromise` for this

danarad05 replied:
Fixed. thanks

amiramw commented:
@vince-fugnitto @RomanNikitenko we have hard time reproducing the linux issue as non of us has linux machine. Only mac and windows. And on WSL we can't make electron work.

Is is ok by you if we merge this PR and open another issue for the minor text problem so someone with linux machine can help?

vince-fugnitto commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree.ts:
```diff
@@ -0,0 +1,105 @@
+    private getChildren(workspaceEdit: monaco.languages.WorkspaceEdit, fileContentsMap: Map<string, string>): BulkEditInfoNode[] {
+        let bulkEditInfos: BulkEditInfoNode[] = [];
+        if (workspaceEdit.edits) {
+            bulkEditInfos = workspaceEdit.edits
+                .map(edit =>
+                    (edit && ('resource' in edit) && edit.resource && edit.resource.path) ? edit.resource.path :
+                        ((edit && ('newUri' in edit) && edit.newUri && edit.newUri.path) ? edit.newUri.path : undefined))
+                .filter((path, index, arr) => path && arr.indexOf(path) === index)
+                .map((path: string) => this.createBulkEditInfo(path, new URI(path), fileContentsMap.get(path)))
+                .filter(Boolean);
```
Can we simplify this part? It is difficult to read and follow and will lead to maintainability issues.

vince-fugnitto commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree.ts:
```diff
@@ -0,0 +1,105 @@
+                .map(edit =>
+                    (edit && ('resource' in edit) && edit.resource && edit.resource.path) ? edit.resource.path :
+                        ((edit && ('newUri' in edit) && edit.newUri && edit.newUri.path) ? edit.newUri.path : undefined))
+                .filter((path, index, arr) => path && arr.indexOf(path) === index)
```
Won't this always be `true`?
```ts
arr.indexOf(path) === index)
```

danarad05 replied:
this line creates distinct paths:
![image](https://user-images.githubusercontent.com/71069170/99272393-6c924a80-2830-11eb-8047-d1661e19190a.png)

danarad05 replied:
@vince-fugnitto Updated to:
private getResourcePath(edit: monaco.languages.WorkspaceTextEdit | monaco.languages.WorkspaceFileEdit): string | undefined {
        return WorkspaceTextEdit.is(edit) ? edit.resource.path : WorkspaceFileEdit.is(edit) && edit.newUri ? edit.newUri.path : undefined;
    }

vince-fugnitto commented on packages/core/src/browser/navigatable.ts:
```diff
@@ -68,7 +68,8 @@ export namespace NavigatableWidget {
     kind: 'navigatable',
-    uri: string
+    uri: string,
+    contents: string | undefined;
```
`contents?: string` ?

danarad05 replied:
Fixed

vince-fugnitto commented on packages/core/src/browser/quick-view-service.ts:
```diff
@@ -28,11 +28,9 @@ export interface QuickViewItem {
 @injectable()
 export class QuickViewService implements QuickOpenModel, QuickOpenHandler, QuickOpenContribution {
-
+    private hiddenItems: (QuickOpenItem & { when?: string })[] = [];
```
Can you explain what the `hiddenItems` are used for?

danarad05 replied:
As per @RomanNikitenko code review, I added functionality for hiding/showing items from Quick Open view. Currently if you remove items from quick open list you cannot return them and you do not have original list saved - you need to recreate entire list which I didn't like as a solution - I thought this solution is least intrusive. 

> **_6. VS Code allows to display Refactor Preview panel if refactoring is in progress, I'm not able to display the panel after canceling refactoring for example. Theia displays empty panel with the corresponding message:_**
> 
> > a) Current system functionality (as far as I can tell) is disposing of widget on closing of tab (on X button) - so I cannot display workspaceEdits after it.
> > b) I fixed so that "Refactor Preview" in displayed in quick open view only when "Refactor Preview" bottom pane is displayed. otherwise it is hidden from quick open view.
> > When Refactor Preview pane is displayed:
> > ![image](https://user-images.githubusercontent.com/71069170/97596877-decff600-1a0d-11eb-86bf-84c7ad69b1d3.png)
> 
> When Refactor Preview pane is not displayed:
> ![image](https://user-images.githubusercontent.com/71069170/97596841-d37cca80-1a0d-11eb-913b-a494c7bc8fca.png)

vince-fugnitto commented on packages/editor/src/browser/editor-manager.ts:
```diff
@@ -138,7 +138,7 @@ export class EditorManager extends NavigatableWidgetOpenHandler<EditorWidget> {
     async open(uri: URI, options?: EditorOpenerOptions): Promise<EditorWidget> {
         const editor = await super.open(uri, options);
-        this.revealSelection(editor, options, uri);
+        setTimeout(() => this.revealSelection(editor, options, uri));
```
Can you explain why this is necessary? If `milliseconds` is omitted, the timeout is `0`.

danarad05 replied:
Before this change, when selecting line in Refactor Preview pane, first time the file is opened in new tab, relevant line is not selected. That seemed to me like the less intrusive fix for this issue.

![7989](https://user-images.githubusercontent.com/71069170/99269336-eb868380-282e-11eb-9069-ce61b8b191ed.gif)

akosyakov replied:
You mean `setTimeout` guruantees that an editor in the proper condition to reveal selection. Please revert.

danarad05 replied:
Reverted

vince-fugnitto commented on packages/plugin-dev/compile.tsconfig.json:
```diff
@@ -25,14 +25,14 @@
-    {
-      "path": "../plugin-ext/compile.tsconfig.json"
-    },
     {
       "path": "../preferences/compile.tsconfig.json"
     },
     {
       "path": "../workspace/compile.tsconfig.json"
+    },
+    {
+      "path": "../plugin-ext/compile.tsconfig.json"
```
The changes are unnecessary.

danarad05 replied:
Fixed

vince-fugnitto commented on packages/plugin-ext-vscode/compile.tsconfig.json:
```diff
@@ -26,14 +26,14 @@
     },
     {
       "path": "../userstorage/compile.tsconfig.json"
+    },
+    {
+      "path": "../plugin-ext/compile.tsconfig.json"
```
The changes are unnecessary.

danarad05 replied:
Fixed

danarad05 commented on examples/browser/package.json:
```diff
@@ -14,6 +14,7 @@
     }
   },
   "dependencies": {
+    "@theia/bulk-edit": "^1.7.0",
     "@theia/api-samples": "^1.7.0",
```
Fixed

danarad05 commented on examples/electron/package.json:
```diff
@@ -14,6 +14,7 @@
     }
   },
   "dependencies": {
+    "@theia/bulk-edit": "^1.7.0",
     "@theia/api-samples": "^1.7.0",
```
Fixed

danarad05 commented on packages/bulk-edit/compile.tsconfig.json:
```diff
@@ -0,0 +1,28 @@
+    "src"
+  ],
+  "references": [
+    {
+      "path": "../core/compile.tsconfig.json"
+    },
+    {
+      "path": "../editor/compile.tsconfig.json"
+    },
+    {
+      "path": "../monaco/compile.tsconfig.json"
+    },
+    {
+      "path": "../filesystem/compile.tsconfig.json"
+    },
+    {
+      "path": "../workspace/compile.tsconfig.json"
+    }
```
Fixed

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree.spec.ts:
```diff
@@ -0,0 +1,83 @@
+/********************************************************************************
+ * Copyright (C) 2017 TypeFox and others.
```
Fix

danarad05 commented:
@vince-fugnitto 
Thank you Vince for review. Please review fixes and remarks. Thanks

vince-fugnitto commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree.ts:
```diff
@@ -0,0 +1,109 @@
+        };
+    }
+
+    private getResourcePath(edit: monaco.languages.WorkspaceTextEdit | monaco.languages.WorkspaceFileEdit): string | undefined {
+        return (edit && ('resource' in edit) && edit.resource && edit.resource.path) ?
+            edit.resource.path :
+            ((edit && ('newUri' in edit) && edit.newUri && edit.newUri.path) ? edit.newUri.path : undefined);
+    }
+}
```
I do not believe `edit` can be `undefined`, there's no need to check for it with this method signature.
We also have the following two methods which can be used:

https://github.com/eclipse-theia/theia/blob/cd85908d3bb5af06477ead7168110f2306a42145/packages/monaco/src/browser/monaco-workspace.ts#L33-L49

danarad05 replied:
Fixed

vince-fugnitto commented:
> @vince-fugnitto @RomanNikitenko we have hard time reproducing the linux issue as non of us has linux machine. Only mac and windows. And on WSL we can't make electron work.
> 
> Is is ok by you if we merge this PR and open another issue for the minor text problem so someone with linux machine can help?

@amiramw @danarad05 I believe it will need to be addressed, currently there is a bug where the input box does not respect the keybinding (irregardless of operating system) which can be updated in the `keymaps.json`. The input displays a hardcoded keybinding rather than displaying the actual value from the application (possibly the reason for the bug on Linux), while in vscode handling is done to update the label.

- [vscode rename update](https://github.com/microsoft/vscode/blob/8134d8c9e7ca00b43ec7ec36034c4dd345eed36b/src/vs/editor/contrib/rename/renameInputField.ts#L85).


![rename](https://user-images.githubusercontent.com/40359487/99285896-258f5f80-2806-11eb-84e2-1292a736ea89.gif)

akosyakov commented on packages/core/src/common/uri.ts:
```diff
@@ -18,16 +18,17 @@ import { URI as Uri } from 'vscode-uri';
-
     private readonly codeUri: Uri;
     private _path: Path | undefined;
+    private _contents: string | undefined;
```
why do we need this? please revert

danarad05 replied:
@akosyakov 
cc: @vince-fugnitto @amiramw 

1. Thanks for review.
2. This was done after @RomanNikitenko found performance issue - which doesn't really relate to this PR (as you can reproduce it using search & replace) - but I thought I would try to solve it. 

The issue this want to address is this: 
When doing search &replace for instance, we are using URI withQuery to save file contents in URI query 

![image](https://user-images.githubusercontent.com/71069170/99366259-09042d80-28c1-11eb-9b4d-9e98cd51ac58.png)

The larger the file is - the longer it takes to display it - for instance opening this search/replace file took ~5-6 seconds and that is mainly because of "percentDecode" function which decodes entire uri querystring.
![image](https://user-images.githubusercontent.com/71069170/99366330-233e0b80-28c1-11eb-82c6-88285fcf8c00.png)

Please advise
Thanks

danarad05 replied:
@akosyakov @vince-fugnitto 
cc: @amiramw 
Reverted this changes - removed "contents" from uri.

akosyakov commented on packages/core/src/browser/diff-uris.ts:
```diff
@@ -23,28 +23,31 @@ export namespace DiffUris {
-        const diffUriStr = JSON.stringify(diffUris);
-
-        return new URI().withScheme(DIFF_SCHEME).withPath(label || '').withQuery(diffUriStr);
+        return new URI().withScheme(DIFF_SCHEME).withPath(label || '').withQuery(diffUris).withContents(diffUriContents);
```
get rid of `withContents` it is not a part of URI spec, but something which this PR invented to fix a particular issue

danarad05 replied:
Please see https://github.com/eclipse-theia/theia/pull/8589#discussion_r524977010

danarad05 replied:
@akosyakov @vince-fugnitto 
cc: @amiramw 
Reverted this changes - removed "contents" from uri.

akosyakov commented on packages/core/src/browser/navigatable.ts:
```diff
@@ -82,7 +83,8 @@ export abstract class NavigatableWidgetOpenHandler<W extends NavigatableWidget>
             kind: 'navigatable',
-            uri: this.serializeUri(uri)
+            uri: this.serializeUri(uri),
+            contents: uri.contents
```
why contents is a part of options which unique identiry it?

danarad05 replied:
Please see https://github.com/eclipse-theia/theia/pull/8589#discussion_r524977010

danarad05 replied:
@akosyakov @vince-fugnitto 
cc: @amiramw 
Reverted this changes - removed "contents" from uri.

akosyakov reviewed (dismissed):
important core APIs like URI and behaviour like of EditorManager is changed unnecessary and bogusly. Please revert these files and try to find more appropriate solution not requiring such changes.

danarad05 commented:
> @amiramw @danarad05 I believe it will need to be addressed, currently there is a bug where the input box does not respect the keybinding (irregardless of operating system) which can be updated in the `keymaps.json`. The input displays a hardcoded keybinding rather than displaying the actual value from the application (possibly the reason for the bug on Linux), while in vscode handling is done to update the label.

@vince-fugnitto The reason that user settings are not overriding default settings is because [dynamicKeyBindings](https://github.com/microsoft/vscode/blob/cd3caa2d7d7379bd1ae6e43bb38de1e953bc1ad4/src/vs/editor/standalone/browser/simpleServices.ts#L365) are not set.
I think this should be handled in a different PR.

![image](https://user-images.githubusercontent.com/71069170/99404049-d7588a00-28f3-11eb-8f12-4afbf85a1983.png)

danarad05 commented:
@vince-fugnitto @akosyakov 
cc: @amiramw 
1. I reverted all "contents" changes back - now content is in uri query. We should probably open an issue concerning performance issue of usage of uri query.
2. We are left we 2 issues currently (as far as I can tell):
a) On Linux electron - malformatted label - we are unable to fix this as we have no Linux machine currently - IMO we should create separate issue for this as this is IMO a minor issue.
b) User keybindings are not overriding default ones in vscode editor (we need to pass them)- this is an existing issue and not related to this PR so should be address in new PR IMO.

Awaiting your response,
Thanks.

danarad05 commented:
@vince-fugnitto 
cc: @amiramw 
Vince, finally managed to reproduce label formatting issue on Linux electron and pushed a fix for it (packages/core/src/browser/keyboard/keyboard-layout-service.ts).
Please review again.
Thanks.
-Dan

vince-fugnitto commented on packages/bulk-edit/package.json:
```diff
@@ -0,0 +1,49 @@
+  "version": "1.7.0",
+  "description": "Theia - Bulk Edit Extension",
+  "dependencies": {
+    "@theia/core": "^1.7.0",
+    "@theia/editor": "^1.7.0",
+    "@theia/filesystem": "^1.7.0",
+    "@theia/monaco": "^1.7.0",
+    "@theia/workspace": "^1.7.0"
```
Please be sure to also update the following dependencies to `^1.8.0`.
The build currently does not succeed due to it (verified locally).

danarad05 replied:
Sorry about that. must have not saved file before pushing.  Fixed

danarad05 commented:
@vince-fugnitto 
Saw that build failed with "UnhandledPromiseRejectionWarning: ReferenceError: monaco is not defined" error. Not sure how to solve this issue. Any suggestions?
Thanks

vince-fugnitto commented:
> @vince-fugnitto
> Saw that build failed with "UnhandledPromiseRejectionWarning: ReferenceError: monaco is not defined" error. Not sure how to solve this issue. Any suggestions?

@danarad05 I have not checked but even performing the tests on the `@theia/bulk-edit` extension fails for me:

```
$ npx run test @theia/bulk-edit
...
bulk-edit-tree
    1) initialize tree


  0 passing (14ms)
  1 failing

  1) bulk-edit-tree
       initialize tree:
     TypeError: Cannot read property 'children' of undefined
      at Context.<anonymous> (lib/browser/bulk-edit-tree/bulk-edit-tree.spec.js:80:34)
      at processImmediate (internal/timers.js:456:21)
```

danarad05 commented:
@vince-fugnitto 
https://github.com/eclipse-theia/theia/pull/8589#issuecomment-736635594
yes that's what I said - it fails first on this error "ReferenceError: monaco is not defined".

danarad05 commented:
@vince-fugnitto fixed test issue and rebased. Please review again.

vince-fugnitto commented on packages/monaco/src/browser/style/index.css:
```diff
@@ -104,3 +104,7 @@
     outline-color: var(--theia-focusBorder);
 }
+
+.monaco-editor .rename-box.preview input {
+    color: var(--theia-editor-foreground);
+}
```
@danarad05 can you update the standard `rename` as well, currently when the preference is off the display is incorrect:

![rename-pref-off](https://user-images.githubusercontent.com/40359487/100891828-aea1ca00-3487-11eb-93c1-45fa7f0990bd.gif)

danarad05 replied:
@vince-fugnitto 
Fixed. Please review again. 
Thanks

vince-fugnitto reviewed (commented):
I verified the functionality of the changes and it works really well :+1: 
It would be good to have other reviewers verify the functionality and code.

danarad05 commented:
@benoitf @akosyakov @vinokurig @amiramw @marechal-p 
As per @vince-fugnitto remark - any additional review would be appreciated.

Thanks,
-Dan

danarad05 commented:
> @benoitf @akosyakov @vinokurig @amiramw @marechal-p
> As per @vince-fugnitto remark - any additional review would be appreciated.
> 
> Thanks,
> -Dan

reminder

amiramw commented:
@danarad05 can u rebase, resolve conflicts and change every place with 1.8.0 to 1.9.0?

danarad05 commented:
> @danarad05 can u rebase, resolve conflicts and change every place with 1.8.0 to 1.9.0?
@amiramw Done
Could you rerun ubuntu-18.04, Node.js v12.x, Python 2.x build.
Thanks

amiramw commented:
When I test rename with F2 I see strikeout also on the added text.

Theia (Gitpod):
![image](https://user-images.githubusercontent.com/9420921/104854886-ebc7fb80-5911-11eb-9d80-352048049ced.png)

VS Code:
![image](https://user-images.githubusercontent.com/9420921/104854913-0ef2ab00-5912-11eb-9632-ce7d867932a8.png)

amiramw commented:
Sometimes it take a lot of time for the refactoring to calculate. For example try to rename `EditorManager`. It took me ~30 seconds.
It would be nice to have a progress indication on such long operations.

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-commands.ts:
```diff
@@ -0,0 +1,33 @@
+/********************************************************************************
+ * Copyright (c) 2020 SAP SE or an SAP affiliate company and others.
```
done

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-contribution.ts:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    private apply(): void {
+        if (this.workspaceEdit && this.workspaceEdit.edits) {
```
done

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-frontend-module.ts:
```diff
@@ -0,0 +1,44 @@
+import { ContainerModule } from 'inversify';
+import { WidgetFactory } from '@theia/core/lib/browser/widget-manager';
+import { BulkEditTreeWidget, BULK_EDIT_TREE_WIDGET_ID, createBulkEditTreeWidget, InMemoryTextResourceResolver } from './bulk-edit-tree';
+import { FrontendApplicationContribution, LabelProviderContribution, } from '@theia/core/lib/browser';
```
done

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree-label-provider.ts:
```diff
@@ -0,0 +1,71 @@
+
+    canHandle(element: object): number {
+        return BulkEditInfoNode.is(element) ?
+            this.treeLabelProvider.canHandle(element) + 1 :
```
@amiramw 
I did same as in:
packages\markers\src\browser\marker-tree-label-provider.ts 
packages\filesystem\src\browser\file-tree\file-tree-label-provider.ts

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree-label-provider.ts:
```diff
@@ -0,0 +1,71 @@
+            description.push(this.labelProvider.getName(rootUri));
+        }
+        // If the given resource is not at the workspace root, include the parent directory to the label.
+        if (rootUri && rootUri.toString() !== node.uri.parent.toString()) {
```
done (did similar to different location)

danarad05 commented on packages/core/src/browser/keyboard/keyboard-layout-service.ts:
```diff
@@ -99,7 +99,7 @@ export class KeyboardLayoutService {
         if (layout) {
             const value = layout.code2Character[key.code];
-            if (value) {
+            if (value && value.replace(/[\n\r\t]/g, '')) {
```
This was a fix for a bug where code2Character map contained newline/tabs as value for the relevant keycodes which resulted in a problem (that I don't recall exactly - maybe it wasn't displayed properly in preferences)

danarad05 commented on packages/editor/src/browser/editor-preferences.ts:
```diff
@@ -805,6 +805,11 @@ const codeEditorPreferenceProperties = {
         'minimum': 0,
         'maximum': 1073741824
     },
+    'editor.rename.enablePreview': {
```
It is a preference - as exists also on VS Code.
![image](https://user-images.githubusercontent.com/71069170/104909826-1e650900-5991-11eb-90d8-70dd00312969.png)

amiramw replied:
where is it used in the code?

danarad05 replied:
It's used in vscode for the "rename" input field
![image](https://user-images.githubusercontent.com/71069170/105624399-0c201a80-5e2a-11eb-9bd2-4c0aa4aa3ebd.png)

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree/in-memory-text-resource.ts:
```diff
@@ -0,0 +1,41 @@
+
+export const MEMORY_TEXT = 'mem-txt';
+
+export class InMemoryTextResource implements Resource {
```
Added some comments. Didn't want to clutter too much - as I based on other Resource implementations that don't have currently any comments - like

danarad05 commented on packages/plugin-ext/src/main/browser/languages-main.ts:
```diff
@@ -1017,9 +1017,15 @@ export function toMonacoWorkspaceEdit(data: WorkspaceEditDto | undefined): monac
         edits: (data && data.edits || []).map(edit => {
             if (WorkspaceTextEditDto.is(edit)) {
-                return { resource: monaco.Uri.revive(edit.resource), edit: edit.edit };
+                return <monaco.languages.WorkspaceTextEdit>{
```
Yes, without it getting ts error
![image](https://user-images.githubusercontent.com/71069170/104911378-71d85680-5993-11eb-9448-3f77ca5f9e15.png)

amiramw replied:
It seems that the reason to the compilation error is that `WorkspaceEditMetadata` and `WorkspaceEditMetadataDto` are not complient. One of them use UriComponents and the other URIs. 

https://github.com/eclipse-theia/theia/blob/5a07bb56c0cade071ee9b767c0040a59d7cdb032/packages/plugin-ext/src/common/plugin-api-rpc-model.ts#L352-L362

https://github.com/eclipse-theia/theia/blob/5a07bb56c0cade071ee9b767c0040a59d7cdb032/packages/plugin-ext/src/common/plugin-api-rpc.ts#L1177-L1187

What is the reason of this difference?

danarad05 replied:
This is similar implementation in Theia as in VSCode (not my code):
WorkspaceEditMetadataDto and other Dto interfaces are used in conjunction with extension/plugin communications while non-Dto interfaces are used inside Theia/Vscode.

From Vscode:
![image](https://user-images.githubusercontent.com/71069170/105684382-8d94ad00-5efd-11eb-8773-5effe319b9b8.png)

![image](https://user-images.githubusercontent.com/71069170/105684443-a3a26d80-5efd-11eb-8d26-b6f2a21527d2.png)

Therefore `fromWorkspaceEdit` function is doing conversion from theia.WorkspaceEdit to WorkspaceEditDto.

danarad05 commented:
> When I test rename with F2 I see strikeout also on the added text.

Fixed

danarad05 commented on packages/core/src/browser/quick-view-service.ts:
```diff
@@ -28,11 +28,9 @@ export interface QuickViewItem {
 @injectable()
 export class QuickViewService implements QuickOpenModel, QuickOpenHandler, QuickOpenContribution {
-
+    private hiddenItems: (QuickOpenItem & { when?: string })[] = [];
```
1. I wanted to have same signature as the items array - in order to "hide" items in "hiddenItems" array 

```
hideItem(label: string): void {
        const index = this.items.findIndex(item => (item.getLabel() === label));
        if (index !== -1) {
            this.hiddenItems = this.hiddenItems.concat(this.items.splice(index, 1));
        }
    }
```
2. Added minimal documentation. It in essence is a solution for an reject Roman had in comparison to VS Code.

amiramw replied:
I suggest to achieve the same hide functionality with implementing `hiddenItems` as a `Set` of labels. Why does it have to have the same type as the items? The usage is different, it only cares about the labels and use this array as a label set actually.

danarad05 replied:
You are right - my bad - it can be simplified as you suggested.
Fixed

danarad05 commented:
> Sometimes it take a lot of time for the refactoring to calculate. For example try to rename `EditorManager`. It took me ~30 seconds.
> Please check if there isn't a bug causing it to take so much time. Otherwise, it would be nice to have a progress indication on such long operations.

1. You are referring to just running Rename refactor and not opening a file from refactor preview (which is a know issue #8768), correct?
2. Did you get 30 seconds locally or on gitpod?
3. I tried renaming EditorManager class and it took around 3 seconds which resulted in 62 files changed in Refactor Preview. In those 3 seconds, most of the time was reading contents of files changed for display in Refactor Preview - i.e. **getFileContentsMap** function in **bulk-edit-tree-widget.tsx**

I did an improvement there which improved time by second+.
Before improvement:
![image](https://user-images.githubusercontent.com/71069170/104918204-18752500-599d-11eb-8030-802bdd15de30.png)
After improvement:
![image](https://user-images.githubusercontent.com/71069170/104920441-735c4b80-59a0-11eb-9341-ed460417a18b.png)

4. I didn't manage to reproduce 30 seconds. If you can elaborate how to reproduce it I could check it again.

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx:
```diff
@@ -0,0 +1,235 @@
+@injectable()
+export class BulkEditTreeWidget extends TreeWidget {
+    private editorWidgets: EditorWidget[] = [];
+    protected readonly toDisposeOnCurrentWidgetChanged = new DisposableCollection();
```
fixed

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx:
```diff
@@ -0,0 +1,235 @@
+    private async getFileContentsMap(workspaceEdit: monaco.languages.WorkspaceEdit): Promise<Map<string, string>> {
+        const fileContentMap = new Map<string, string>();
+
+        if (workspaceEdit && workspaceEdit.edits) {
```
done

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx:
```diff
@@ -0,0 +1,235 @@
+            let resource;
+            for (const element of workspaceEdit.edits) {
+                if (element) {
+                    const filePath = (('newUri' in element) && element.newUri && element.newUri.path) ? element.newUri.path :
```
done

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx:
```diff
@@ -0,0 +1,235 @@
+        const fileContentMap = new Map<string, string>();
+
+        if (workspaceEdit && workspaceEdit.edits) {
+            let fileUri;
```
done

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx:
```diff
@@ -0,0 +1,235 @@
+    }
+
+    protected decorateBulkEditNode(node: BulkEditNode): React.ReactNode {
+        if (node && node.parent && node.bulkEdit && ('edit' in node.bulkEdit)) {
```
done

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx:
```diff
@@ -0,0 +1,235 @@
+            const bulkEdit = node.bulkEdit;
+            const parent = node.parent as BulkEditInfoNode;
+
+            if (parent && parent.fileContents) {
```
done

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx:
```diff
@@ -0,0 +1,235 @@
+            const parent = node.parent as BulkEditInfoNode;
+
+            if (parent && parent.fileContents) {
+                const lines = parent.fileContents.split('\n');
```
contents is joined with "\n" (as similarily is done also at `packages\search-in-workspace\src\browser\search-in-workspace-result-tree-widget.tsx`)
![image](https://user-images.githubusercontent.com/71069170/105152865-68282d80-5b10-11eb-9a32-02cd7edd72d0.png)

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx:
```diff
@@ -0,0 +1,235 @@
+                    const startColumn = +bulkEdit.edit.range.startColumn;
+                    const endColumn = +bulkEdit.edit.range.endColumn;
+                    const lineText = lines[startLineNum - 1];
+                    const beforeMatch = (startColumn > 26 ? '... ' : '') + lineText.substr(0, startColumn - 1).substr(-25);
```
I did it similarily to logic in search/replace results:
![image](https://user-images.githubusercontent.com/71069170/105153525-36fc2d00-5b11-11eb-94ce-8ae4862ae065.png)

If position where change starts in from than 2 chars from start of line that it adds "..." in begining

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx:
```diff
@@ -0,0 +1,235 @@
+                    const lineText = lines[startLineNum - 1];
+                    const beforeMatch = (startColumn > 26 ? '... ' : '') + lineText.substr(0, startColumn - 1).substr(-25);
+                    const replacedText = lineText.substring(startColumn - 1, endColumn - 1);
+                    const afterMatch = lineText.substr(startColumn - 1 + replacedText.length, 75);
```
Same as above - similar logic implemented as in search and replace - 75 is max amount of chars to display for each line in refactor preview changes

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx:
```diff
@@ -0,0 +1,235 @@
+    private async createReplacePreview(bulkEditInfoNode: BulkEditInfoNode): Promise<URI> {
+        const fileUri = bulkEditInfoNode.uri;
+        let lines: string[] = [];
+        if (bulkEditInfoNode && bulkEditInfoNode.fileContents) {
```
done

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx:
```diff
@@ -0,0 +1,235 @@
+        const fileUri = bulkEditInfoNode.uri;
+        let lines: string[] = [];
+        if (bulkEditInfoNode && bulkEditInfoNode.fileContents) {
+            lines = bulkEditInfoNode.fileContents.split('\n');
```
see https://github.com/eclipse-theia/theia/pull/8589#discussion_r560799217

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx:
```diff
@@ -0,0 +1,235 @@
+
+    private getEditorOptions(node: BulkEditNode): EditorOpenerOptions {
+        let options = {};
+        if (('edit' in node.bulkEdit) && node.bulkEdit.edit && node.bulkEdit.edit.range) {
```
done

danarad05 commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree.ts:
```diff
@@ -0,0 +1,108 @@
+            if (bulkEditInfos.length > 0) {
+                bulkEditInfos.forEach(editInfo => {
+                    editInfo.children = workspaceEdit.edits.filter(edit => ((('resource' in edit) && edit.resource.path === editInfo.id)) ||
+                        (('newUri' in edit) && edit.newUri && edit.newUri.path === editInfo.id))
```
done

amiramw commented:
> 1. You are referring to just running Rename refactor and not opening a file from refactor preview (which is a know issue #8768), correct?

Yes. Calculating list of changes to be displayed in the new view.

> 2. Did you get 30 seconds locally or on gitpod?

Gitpod. 

> 3. I tried renaming EditorManager class and it took around 3 seconds which resulted in 62 files changed in Refactor Preview. In those 3 seconds, most of the time was reading contents of files changed for display in Refactor Preview - i.e. **getFileContentsMap** function in **bulk-edit-tree-widget.tsx**
> 
> I did an improvement there which improved time by second+.
> Before improvement:
> ![image](https://user-images.githubusercontent.com/71069170/104918204-18752500-599d-11eb-8030-802bdd15de30.png)
> After improvement:
> ![image](https://user-images.githubusercontent.com/71069170/104920441-735c4b80-59a0-11eb-9341-ed460417a18b.png)
> 
> 1. I didn't manage to reproduce 30 seconds. If you can elaborate how to reproduce it I could check it again.

Now it takes 10 seconds in gitpod. I guess it is good enough for big refactoring.

amiramw commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx:
```diff
@@ -0,0 +1,230 @@
+        if (workspaceEdit?.edits) {
+            for (const element of workspaceEdit.edits) {
+                if (element) {
+                    const filePath = (('newUri' in element) && element?.newUri?.path) || (('resource' in element) && element?.resource?.path);
```
do you need `('newUri' in element)` and `('resource' in element)` conditions?

danarad05 replied:
yes.  A workspaceEdit can be Text of File and each has different props and therefore the check is needed
![image](https://user-images.githubusercontent.com/71069170/105686908-7c00d480-5f00-11eb-9fed-11a101f8d0b5.png)

danarad05 commented:
@amiramw @azatsarynnyy Please review again - if issue in conversation is satisfactory please resolve conversation.

amiramw commented on packages/core/src/browser/quick-view-service.ts:
```diff
@@ -28,11 +28,10 @@ export interface QuickViewItem {
 export class QuickViewService implements QuickOpenModel, QuickOpenHandler, QuickOpenContribution {
-
+    // hiddenItemLabels contains item labels hidden from Quick View
+    private hiddenItemLabels = new Set<string | undefined>();
```
i think it should not hold undefined. maybe instead add null check on line 84?

danarad05 commented on packages/core/src/browser/quick-view-service.ts:
```diff
@@ -59,6 +59,14 @@ export class QuickViewService implements QuickOpenModel, QuickOpenHandler, Quick
         });
     }
 
+    hideItem(label: string): void {
+        this.hiddenItemLabels.add(label);
+    }
+
+    showItem(label: string): void {
+        this.hiddenItemLabels.delete(label);
+    }
+
```
The set and get do not allow undefined so therefore actually `hiddenItemLabels` will never contain undefined.. so I thought it is cleaner solution.

vince-fugnitto commented on packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx:
```diff
@@ -0,0 +1,230 @@
+            lines = bulkEditInfoNode.fileContents.split('\n');
+            bulkEditInfoNode.children.map((node: BulkEditNode) => {
+                if (node.bulkEdit && ('edit' in node.bulkEdit)) {
+                    const startLineNum = +node.bulkEdit.edit.range.startLineNumber;
+                    if (lines.length > startLineNum) {
+                        const startColumn = +node.bulkEdit.edit.range.startColumn;
+                        const endColumn = +node.bulkEdit.edit.range.endColumn;
```
Minor: why is the `+` cast necessary? `startLineNumber`, `startColumn` and `endColumn` are numbers already?
In addition, we are not doing it in other areas such as:

```ts
selection: {
    start: {
        line: node.bulkEdit.edit.range.startLineNumber - 1,
        character: node.bulkEdit.edit.range.startColumn - 1
    },
    end: {
        line: node.bulkEdit.edit.range.endLineNumber - 1,
        character: node.bulkEdit.edit.range.endColumn - 1
    }
}
```

danarad05 replied:
Fixed. Thanks

amiramw reviewed (approved):
LGTM, just rebase and fix versions to 1.10.0

danarad05 commented:
> LGTM, just rebase and fix versions to 1.10.0

@amiramw Done

amiramw commented:
@akosyakov this PR can't be merged due to your requested changes. I believe they were addressed by @danarad05 , can you take a look?

amiramw commented:
@vince-fugnitto do you think we can merge this?

vince-fugnitto commented:
> @vince-fugnitto do you think we can merge this?

@amiramw I'll leave it up to your discretion, if the pull-request required CQ (since we copy or are inspired by the bulk-edit implementation of vscode), can you confirm it's been approved. Generally a link in the pull-request description is added for other committers to track.

amiramw commented:
> @amiramw I'll leave it up to your discretion, if the pull-request required CQ (since we copy or are inspired by the bulk-edit implementation of vscode), can you confirm it's been approved. Generally a link in the pull-request description is added for other committers to track.

Good catch. Created now CQ https://dev.eclipse.org/ipzilla/show_bug.cgi?id=23016

amiramw commented:
According the CQ:
> We have completed a high level triage of the attachment and you may now check
the content into your project’s source code repository. If a repository has not
yet been allocated, please connect with webmaster@eclipse.org to either create
a new repository or move your existing repository to the Eclipse Foundation. 
We will perform a comprehensive analysis at a later date based on IP work
queue.

@vince-fugnitto this means that we can merge this PR, right?

vince-fugnitto commented:
> @vince-fugnitto this means that we can merge this PR, right?

@amiramw correct :)

amiramw commented:
Merged. Thanks everyone for reviewing.
